### DISCUSSION TITLE
fileFilter at processPending uploads all even if response is false

### DISCUSSION THREAD

**Discussion by OscarEstudilloMarin** [rpldy/react-uploady#771] [Q&A] [answered]
fileFilter at processPending uploads all files (with autoUpload false) even if response is false

const uploady = useUploady();

uploady.processPending({
	fileFilter: () => {
		return false
	}
});

**Expected behavior**
None of the files are uploaded (false is just for testing, on the real scenario I want to make some conditions in order to decide which of the files are going to be uploaded)

**Comment by yoavniran** [comment]
Hey @OscarEstudilloMarin 
Thinking about this for a second more, I realized the issue here. 
When calling `processPending`, the files were already added and therefore will not be filtered. Uploady doesn't re-run the filter at this point (and Im not sure it makes sense to do so).

So, beyond documenting better this situation (there are other params that arent applied at this point), can you elaborate why you're setting the fileFilter at the point of calling `processPending` and not earlier (ex: as a prop on the Uploady component)?

**Comment by OscarEstudilloMarin** [comment]
Hi @yoavniran,

I'd like to explain why fileFilter in processPending is essential for our use case:

In our application, we have a partner role that can perform various actions for different clients, including file uploads. Rather than implementing automatic uploads, we've designed a two-step process that allows for pre-upload actions.

Here's how it works:
1. Partners can queue files for multiple clients
2. The actual upload only begins when the partner clicks "Start Upload" (which triggers processPending)

The expected behavior varies depending on the context:
- From the general uploads page: All queued files for all clients will upload
- From a specific client's page: Only files queued for that particular client will upload (this is what it's not working due to fileFilter not filtering at processPending)

I've attached a video demonstration to better illustrate this workflow.

https://github.com/user-attachments/assets/eb3c2a0a-9c8f-414a-9181-c088f5ca5aa5

**Comment by yoavniran** [comment, marked as answer]
Thanks @OscarEstudilloMarin ,
its much clearer now.

I think _fileFilter_ isn't the right tool here. I'd suggest using the "cancellable" event handlers/hooks for this purpose. 
_fileFilter_ was designed to filter out files immediately when they're added and not after they're in the queue.

Here's a [guide](https://react-uploady.org/docs/guides/DynamicallyCancelUploads/) that discusses cancalling uploads.

Additionally, I think cancelling will allow you to build a UI that makes it clear what happened to the item that was added. filtering simply makes the item disappear as if it wasn't added, which may be confusing to the user. 

Alternatively, if you are able to determine whether files should be filtered out when they are added, but from a different location. It is possible to define the _fileFilter_ at the UploadButton level (for example).

This [guide](https://react-uploady.org/docs/guides/DifferentConfiguration/) shows how you can override settings from the upload component.


Hope this helps. For now, I'll move this to Discussions and you can reply there.

**Comment by OscarEstudilloMarin** [comment]
Thanks a lot for the explanation @yoavniran 

I was convinced fileFilter didn't remove the files so that is why I thought of using it.

I'll investigate cancellable events in order to achieve the expected functionality.

Thank you very much for your help and for your kindness creating this library 🙌🏽
